### PR TITLE
Task01 Валерий Миронов CSC

### PR DESCRIPTION
**Pull request by MBkkt**
<details><summary>Локальный вывод</summary><p>

<pre>
Device name: Radeon 500 Series (POLARIS12, DRM 3.40.0, 5.10.64-1-lts, LLVM 12.0.1)
Work Group size: 256
Data generated for n=1000000!
Result data transfer time: 0.00760042+-1.0436e-05 s
RAM -> VRAM bandwidth: 1.47071 GB/s
Kernel average time: 0.00166375+-1.55087e-05 s
GFlops: 0.601167
VRAM bandwidth: 6.71857 GB/s
Result data transfer time: 0.00331792+-6.0757e-05 s
VRAM -> RAM bandwidth: 1.123 GB/s

Device name: Intel(R) UHD Graphics 620 [0x5917]
Work Group size: 256
Data generated for n=1000000!
Result data transfer time: 0.00159383+-8.28088e-05 s
RAM -> VRAM bandwidth: 7.01329 GB/s
Kernel average time: 0.000542667+-9.41925e-06 s
GFlops: 1.84311
VRAM bandwidth: 20.5983 GB/s
Result data transfer time: 0.000265333+-1.41559e-05 s
VRAM -> RAM bandwidth: 14.0427 GB/s

Device name: pthread-Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
Work Group size: 4096
Data generated for n=1000000!
Result data transfer time: 0.000863417+-1.94099e-05 s
RAM -> VRAM bandwidth: 12.9893 GB/s
Kernel average time: 0.000756+-0.000105837 s
GFlops: 1.32741
VRAM bandwidth: 14.8349 GB/s
Result data transfer time: 0.00041075+-2.48734e-05 s
VRAM -> RAM bandwidth: 9.10141 GB/s

Device name: Radeon 500 Series (POLARIS12, DRM 3.40.0, 5.10.64-1-lts, LLVM 12.0.1)
Work Group size: 256
Data generated for n=100000000!
Result data transfer time: 0.111262+-0.000105919 s
RAM -> VRAM bandwidth: 10.0446 GB/s
Kernel average time: 0.528666+-6.66689e-05 s
GFlops: 0.189155
VRAM bandwidth: 2.11398 GB/s
Result data transfer time: 0.389577+-0.0117498 s
VRAM -> RAM bandwidth: 0.956239 GB/s

Device name: Intel(R) UHD Graphics 620 [0x5917]
Work Group size: 256
Data generated for n=100000000!
Result data transfer time: 0.110364+-0.000397155 s
RAM -> VRAM bandwidth: 10.1264 GB/s
Kernel average time: 0.0414528+-0.000118806 s
GFlops: 2.41239
VRAM bandwidth: 26.9605 GB/s
Result data transfer time: 0.0370096+-2.81053e-05 s
VRAM -> RAM bandwidth: 10.0657 GB/s

Device name: pthread-Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
Work Group size: 4096
Data generated for n=100000000!
Result data transfer time: 0.107874+-0.000701167 s
RAM -> VRAM bandwidth: 10.3606 GB/s
Kernel average time: 0.0849187+-0.000106252 s
GFlops: 1.17764
VRAM bandwidth: 13.1612 GB/s
Result data transfer time: 0.0359803+-0.00044472 s
VRAM -> RAM bandwidth: 10.3541 GB/s
</pre>

</p></details>

<details><summary>Вывод Travis CI</summary><p>

<pre>
$ ./build/aplusb
Device name:            Intel(R) Xeon(R) CPU
Work Group size: 8192
Data generated for n=1000000!
Result data transfer time: 0.00328708+-0.000161719 s
RAM -> VRAM bandwidth: 3.42583 GB/s
Log:
Compilation started
Compilation done
Linking started
Linking done
Device build started
Device build done
Kernel <aplusb> was successfully vectorized (8)
Done.
Kernel average time: 0.000542583+-2.40744e-05 s
GFlops: 1.85707
VRAM bandwidth: 20.7544 GB/s
Result data transfer time: 0.000390083+-6.77567e-06 s
VRAM -> RAM bandwidth: 9.62272 GB/s

Device name:            Intel(R) Xeon(R) CPU
Work Group size: 8192
Data generated for n=100000000!
Result data transfer time: 0.430352+-0.00377739 s
RAM -> VRAM bandwidth: 2.59712 GB/s
Log:
Compilation started
Compilation done
Linking started
Linking done
Device build started
Device build done
Kernel <aplusb> was successfully vectorized (8)
Done.
Kernel average time: 0.0911743+-0.000841315 s
GFlops: 1.09689
VRAM bandwidth: 12.2587 GB/s
Result data transfer time: 0.110694+-0.000311689 s
VRAM -> RAM bandwidth: 3.36566 GB/s
The command "./build/aplusb" exited with 0.
</pre>

</p></details>

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -33,145 +30,266 @@ void reportError(cl_int err, const std::string &filename, int line)
...
+            OCL_SAFE_CALL(clEnqueueWriteBuffer(*queue, *bs_gpu, CL_FALSE, 0, sizeof(float) * global_work_size,
+                                               bs.data(), 0, nullptr, nullptr));
+            OCL_SAFE_CALL(clEnqueueWriteBuffer(*queue, *cs_gpu, CL_TRUE, 0, sizeof(float) * global_work_size, cs.data(),
+                                               0, nullptr, nullptr));
+            t.nextLap();
+        }
+        const auto avg_time = t.lapAvg();
+        std::cout << "Result data transfer time: " << avg_time << "+-" << t.lapStd() << " s\n"
+                  << "RAM -> VRAM bandwidth: "
+                  << double(3 * global_work_size * sizeof(float)) / avg_time / (1024 * 1024 * 1024) << " GB/s"
```
Comment (MBkkt) [inline src/main.cpp:205]: Мне не очень нравится что мы меряем вместе разные типы буферов.
Ну да неважно

File: src/main.cpp [hunk 2]
```diff
@@ -33,145 +30,266 @@ void reportError(cl_int err, const std::string &filename, int line)
...
-    //   - Сохранить событие "кернел запущен" (см. аргумент "cl_event *event")
-    //   - Дождаться завершения полунного события - см. в документации подходящий метод среди Event Objects
     {
-        size_t workGroupSize = 128;
-        size_t global_work_size = (n + workGroupSize - 1) / workGroupSize * workGroupSize;
-        timer t; // Это вспомогательный секундомер, он замеряет время своего создания и позволяет усреднять время нескольких замеров
+        timer t;// Это вспомогательный секундомер, он замеряет время своего создания и позволяет усреднять время нескольких замеров
         for (unsigned int i = 0; i < 20; ++i) {
-            // clEnqueueNDRangeKernel...
-            // clWaitForEvents...
```
Comment (MBkkt) [inline src/main.cpp:133]: Я же правильно понял что clEnqueueNDRangeKernel возвращает event который нужно зарелизить(clReleaseEvent)?

Comment (PolarNick239) [reply src/main.cpp:133]: да

File: src/main.cpp [hunk 3]
```diff
@@ -33,145 +30,266 @@ void reportError(cl_int err, const std::string &filename, int line)
...
-    // код по переданному аргументом errcode_ret указателю)
-    // И хорошо бы сразу добавить в конце clReleaseContext (да, не очень RAII, но это лишь пример)
+    ~ClKernel() { OCL_SAFE_CALL(clReleaseKernel(kernel)); }
 
-    // TODO 3 Создайте очередь выполняемых команд в рамках выбранного контекста и устройства
-    // См. документацию https://www.khronos.org/registry/OpenCL/sdk/1.2/docs/man/xhtml/ -> OpenCL Runtime -> Runtime APIs -> Command Queues -> clCreateCommandQueue
-    // Убедитесь, что в соответствии с документацией вы создали in-order очередь задач
-    // И хорошо бы сразу добавить в конце clReleaseQueue (не забывайте освобождать ресурсы)
+private:
+    cl_kernel kernel;
```
Comment (PolarNick239) [inline src/main.cpp:155]: 👍 за RAII

File: src/main.cpp (hunk 2 above)
Comment (PolarNick239) [reply src/main.cpp:133]: спасибо, забыл это проверять :)

**Pull request closed without merging**